### DISCUSSION TITLE
Data Seeding: Create fake image data using image URLs

### DISCUSSION THREAD

**Discussion by rayjasson98** [shrinerb/shrine#581] [Q&A] [answered]
Hi, during data seeding, I want to create fake `image_data` attributes from image URLs. I can get a set of random image URLs from Faker generators like `Faker::LoremFlickr`.  These are my requirements:

1.  Create `image_data` object from image URLs and save it together with the model.
2.  Do not trigger any upload process or backgrounding promote block when seeding fake image data.

According to the docs, I can build an `<attachment>_data` object like this:

```js
  {
    id: "...",
    storage: "...",
    metadata: {
      size: "...",
      filename: "...",
      mime_type: "..."
    }
  }
```
However, the problem is there is no way to create the object from image URLs where I can retrieve the stored image using the `<attachment>_url` method. I know it does not make sense to have this behaviour in production, but is there a way I can simulate this behaviour in order for me to seed image data quickly?

Thank you.

**Comment by jrochkind** [comment]
Having just done something similar, i _think_ what you want is:

    attacher.set(data_hash)

For instance, on an ActiveRecord model with an `image` shrine attribute, 

    model.image_attacher.set( data_hash )

Let me know if it does what you want?

**Reply by rayjasson98** [reply]
Thanks for your hints! Please see my reply below.

**Comment by rayjasson98** [comment, marked as answer]
Unfortunately, calling `model.image_attacher.set(image_data_hash)` gave me this error: 
```
bundle/gems/shrine-3.4.0/lib/shrine/attacher.rb:314:in `file=': expected file to be a Shrine::UploadedFile or nil, got ...
```
While digging into the [source code](https://github.com/shrinerb/shrine/blob/master/lib/shrine/attacher.rb#L312-L315), I found that I need to instantiate a `Shrine::UploadedFile` object with the `image_data_hash`. Also, I need a gem called [shrine-url](https://github.com/shrinerb/shrine-url) to create a fake storage that allows me to assign image data by custom URLs.

The full steps are shown below:

1. Add `gem "shrine-url", "~> 2.4"` to your Gemfile.
2. In your `config/initializers/shrine.rb`, add the following lines:
    ```ruby
    require "shrine/storage/url"
    
    Shrine.storages = {
      test: Shrine::Storage::Url.new,
      ...
    }
    ```
4. Create fake `image_data` attribute like this:
    ```ruby
    image_data = Shrine::UploadedFile.new({
      id: "your image url",
      storage: "test",
      metadata: {}
    })
    ```
6. Save the model with the `image_data` attribute:
    ```ruby
    model.image_attacher.set(image_data)
    mode.save
    ```

You can now retrieve the image by calling `model.image_url`. Also, I think it would be nice if someone can include this in the documentation.

**Reply by jrochkind** [reply]
oops, that's exactly right, I accidentally forgot that uploaded file step, sorry! Another way to do that is:

    model.image_attacher.set(Shrine.uploaded_file(image_data))
